Mathlib/Data/Bitvec/Lemmas.lean

```diff
@@ -1,173 +1,68 @@
-  apply Subtype.ext
-  change Vector.toList _ = xs
-  dsimp [Bitvec.toNat, bitsToNat]
-  rw [← List.length_reverse] at h
-  rw [← List.reverse_reverse xs, List.foldl_reverse]
-  generalize xs.reverse = ys at h ⊢; clear xs
-  induction' ys with ys_head ys_tail ys_ih generalizing n
-  · cases h
-    simp [Bitvec.ofNat]
-  · simp only [← Nat.succ_eq_add_one, List.length] at h
-    subst n
-    simp only [Bitvec.ofNat, Vector.toList_cons, Vector.toList_nil, List.reverse_cons,
-      Vector.toList_append, List.foldr]
-    erw [addLsb_div_two, decide_addLsb_mod_two]
-    congr
-    apply ys_ih
-    rfl
-#align bitvec.of_nat_to_nat Bitvec.ofNat_toNat
-
-theorem toFin_val {n : ℕ} (v : Bitvec n) : (toFin v : ℕ) = v.toNat := by
-  rw [toFin, Fin.coe_ofNat_eq_mod, Nat.mod_eq_of_lt]
-  apply toNat_lt
-#align bitvec.to_fin_val Bitvec.toFin_val
-
-theorem toFin_le_toFin_of_le {n} {v₀ v₁ : Bitvec n} (h : v₀ ≤ v₁) : v₀.toFin ≤ v₁.toFin :=
-  show (v₀.toFin : ℕ) ≤ v₁.toFin by
-    rw [toFin_val, toFin_val]
-    exact h
-#align bitvec.to_fin_le_to_fin_of_le Bitvec.toFin_le_toFin_of_le
-
-theorem ofFin_le_ofFin_of_le {n : ℕ} {i j : Fin (2 ^ n)} (h : i ≤ j) : ofFin i ≤ ofFin j :=
-  show (Bitvec.ofNat n i).toNat ≤ (Bitvec.ofNat n j).toNat by
-    simp only [toNat_ofNat, Nat.mod_eq_of_lt, Fin.is_lt]
-    exact h
-#align bitvec.of_fin_le_of_fin_of_le Bitvec.ofFin_le_ofFin_of_le
-
-theorem toFin_ofFin {n} (i : Fin <| 2 ^ n) : (ofFin i).toFin = i :=
-  Fin.eq_of_veq (by simp [toFin_val, ofFin, toNat_ofNat, Nat.mod_eq_of_lt, i.is_lt])
-#align bitvec.to_fin_of_fin Bitvec.toFin_ofFin
-
-theorem ofFin_toFin {n} (v : Bitvec n) : ofFin (toFin v) = v := by
-  dsimp [ofFin]
-  rw [toFin_val, ofNat_toNat]
-#align bitvec.of_fin_to_fin Bitvec.ofFin_toFin
-
-end Bitvec
+variable {w v : Nat}
+
+@[norm_cast]
+theorem val_bitvec_eq {x y : BitVec w} : x.val = y.val ↔ x = y :=
+  ⟨(match x, y, · with | ⟨_, _⟩,⟨_, _⟩, rfl => rfl), (· ▸ rfl)⟩
+
+/-- `x < y` as natural numbers if and only if `x < y` as `BitVec w`. -/
+@[norm_cast]
+theorem val_bitvec_lt {x y : BitVec w} : x.val < y.val ↔ x < y := by
+  simp [LT.lt, BitVec.lt]
+
+/-- `x ≠ y` as natural numbers if and only if `x != y` as `BitVec w`. -/
+@[norm_cast]
+theorem val_bitvec_bne {x y : BitVec w} : x.val ≠ y.val ↔ x != y := by
+  simp [bne, val_bitvec_eq]
+
+lemma testBit_eq_ofNat {k} {x : BitVec w} :
+  Bool.toNat (testBit (x.val) k) = (BitVec.ofNat 1 (x.val >>> k)).val:= by
+  simp only [BitVec.ofNat, Fin.ofNat', testBit, mod_two_of_bodd, pow_one]
+  aesop
+
+lemma val_to_ofNat {m} (h : m < 2^w) : (BitVec.ofNat w m).val = m := by
+  simp [BitVec.ofNat, Fin.ofNat', mod_eq_of_lt h]
+
+lemma ofNat_to_val (x : BitVec w) : BitVec.ofNat w x.val = x := by
+  simp [BitVec.ofNat, Fin.ofNat', mod_eq_of_lt x.isLt]
+
+lemma ofNat_to_val' (x : BitVec w) (h : v = w):
+  HEq x (BitVec.ofNat v x.val) := h ▸ heq_of_eq (ofNat_to_val x).symm
+
+theorem concat_ext {x : BitVec w} {y : BitVec v} :
```

Comment:
This is about `append` not `concat` right? And it's not an extensionality theorem. It should have a name that has more to do with the statement.